### ISSUE TITLE
Upcoming admin API

### ISSUE THREAD

**Issue by clement911** [nozzlegear/ShopifySharp#227]
Hi @nozzlegear,

Not sure if you've heard that Shopify is working on a GraphQL based API to query the same entities as the 
 REST API? It's already in beta actually.

It would have its place in ShopifySharp right?

**Comment by nozzlegear** [comment]
Yes I have heard that! I've even messed around with the beta implementation they had a couple months ago. I haven't done anything serious with it or thought about adding it to ShopifySharp as I don't have much experience with GraphQL or any .NET GraphQL clients. 

I would love to get it in this package though, or perhaps as its own package like `ShopifySharp.GraphQL`. I wonder how we would go about deserializing the returned data though, since you can just request any arbitrary object you want.

**Comment by nozzlegear** [comment]
I think your last reply is missing, but I think anonymous objects and F# type providers would be a cool solution. If you think of anything else that we could do with the GraphQL API as we get closer to its release I'd love to hear it!

**Comment by nozzlegear** [comment]
Just brainstorming here: in F# it would be possible to use a list of union types to indicate which fields we would want to return from the  GraphQL API. We could then parse the result properties back into a list of union types and the dev could pattern match against the list to get the values.

```fs
type RequestedField =
  | Id
  | CustomerName
  | Address

 type ReturnedField =
  | Id of int64
  | CustomerName of string
  | Address of string

// Tell the lib we want Id and Address
let values = 
  [ 
    RequestedField.Id
    RequestedField.Address
  ]
  |> sendGraphRequest
  
// We now have a list of ReturnedField DUs. 
values
|> List.forall (fun field -> 
  match field with
  | ReturnedField.Id id -> 
    printfn "id value is %i" id
  | ReturnedField.Address add ->
    printfn "address is %s" add
  _ ->
    failwith "Should have only returned id and address"
)
```

Not sure how usable this would be from C#, but I think with C# 7's pattern matching switch statements it might at least be possible.

**Comment by clement911** [comment]
Hmm the thing is GraphQL doesn't return a list, it returns an object graph. 
Or am I missing something?

**Comment by clement911** [comment]
an F# type provider that could infer the shape of the response based on the query string would be epic.

**Comment by nozzlegear** [comment]
Oh yeah, I didn't make it clear. In this case the package itself would take the object returned by GraphQL and parse it into the list of Returned field DUs.  Type providers would of course be a great fit too, but I don't think they're fully supported in .NET Standard right now.

**Comment by clement911** [comment]
After watching a bit of Unite 2018, it's clear that Shopify is serious about GraphQL.
It's a really cool technology actually, I'm glad they are using it.
I still wonder whether it makes sense to augment ShopifySharp to support GraphQL.
The beauty of GraphQL is that it is very simple. There is a single endpoint.
But maybe ShopifySharp could still help to make things even easier.
Looking a https://github.com/bkniffler/graphql-net-client , one can query by passing type: `client.Query(query).Get<T>`
We could potentially define all the entity types in ShopifySharp. But I'm not sure if this is useful because:
-A graphQL query can request partial object => maybe nullable properties might help
-A graphQL query can alias properties
-A graphQL query can be arbitrarily deep
In other words, I don't see a way of making it work except with simple queries.
The other thing that could be useful is handling the authentication and also the retry logic and API rate limit as ShopifySharp does today already...
So I'm still unsure, this may not be worth it. Or maybe it should just be a couple of thin helpers...
